SSA/Projects/InstCombine/LLVM/EDSL.lean

```diff
@@ -1,28 +1,405 @@
 import Qq
+import SSA.Projects.InstCombine.Base
+import Std.Data.BitVec
+import SSA.Projects.MLIRSyntax.AST
 import SSA.Projects.MLIRSyntax.EDSL
 import SSA.Projects.InstCombine.LLVM.Transform
+-- import SSA.Projects.InstCombine.LLVM.Transform.Dialects.InstCombine
 
-open Qq Lean Meta Elab.Term
+open Qq Lean Meta Elab.Term Elab Command
+open InstCombine (MOp MTy Width)
 
+open MLIR
+
+namespace InstcombineTransformDialect
+
+def mkUnaryOp {Γ : Ctxt (MTy φ)} {ty : (MTy φ)} (op : MOp φ)
+  (e : Ctxt.Var Γ ty) : MLIR.AST.ExceptM (MOp φ) <| Expr (MOp φ) Γ ty :=
+ match ty with
+ | .bitvec w =>
+   match op with
+   -- Can't use a single arm, Lean won't write the rhs accordingly
+    | .neg w' => if h : w = w'
+      then return ⟨
+        .neg w',
+        by simp [OpSignature.outTy, signature, h],
+        .cons (h ▸ e) .nil,
+        .nil
+      ⟩
+      else throw <| .widthError w w'
+    | .not w' => if h : w = w'
+      then return ⟨
+        .not w',
+        by simp [OpSignature.outTy, signature, h],
+        .cons (h ▸ e) .nil,
+        .nil
+      ⟩
+      else throw <| .widthError w w'
+    | .copy w' => if h : w = w'
+      then return ⟨
+        .copy w',
+        by simp [OpSignature.outTy, signature, h],
+        .cons (h ▸ e) .nil,
+        .nil
+      ⟩
+      else throw <| .widthError w w'
+    | _ => throw <| .unsupportedUnaryOp
+
+def mkBinOp {Γ : Ctxt (MTy φ)} {ty : (MTy φ)} (op : MOp φ)
+    (e₁ e₂ : Ctxt.Var Γ ty) : MLIR.AST.ExceptM (MOp φ) <| Expr (MOp φ) Γ ty :=
+ match ty with
+ | .bitvec w =>
+   match op with
+   -- Can't use a single arm, Lean won't write the rhs accordingly
+    | .add w' => if h : w = w'
+      then return ⟨
+        .add w',
+        by simp [OpSignature.outTy, signature, h],
+        .cons (h ▸ e₁) <| .cons (h ▸ e₂) .nil ,
+        .nil
+      ⟩
```

Comment:
This looks like it could use some extra abstraction / deduplication. Let's defer that to a separate, low-prio issue